zkevm-circuits/src/evm_circuit/execution.rs

```diff
@@ -1107,19 +1110,16 @@ impl<F: Field> ExecutionConfig<F> {
             }
         }
 
-        for (idx, assigned_rw_value) in assigned_rw_values.iter().enumerate() {
-            let rw_idx = step.rw_indices[idx];
-            let rw = block.rws[rw_idx];
-            let table_assignments = rw.table_assignment(block.randomness);
-            let rlc = table_assignments.rlc(block.randomness);
-            if rlc != assigned_rw_value.1 {
-                log::error!(
-                    "incorrect rw witness. lookup input name: \"{}\". rw: {:?}, rw index: {:?}, {}th rw of step {:?}",
-                    assigned_rw_value.0,
-                    rw,
-                    rw_idx,
-                    idx,
-                    step.execution_state);
+        let rlc_assignments: BTreeSet<_> = block
+            .rws
+            .table_assignments()
+            .iter()
+            .map(|rw| rw.table_assignment(block.randomness).rlc(block.randomness))
+            .collect();
+
+        for (name, value) in assigned_rw_values.iter() {
+            if !rlc_assignments.contains(value) {
+                log::error!("rw lookup error: name: {}, step: {:?}", *name, step);
```

Comment:
it is a bit complex.

(1) this modification is correct. It checks the lookup input is in the lookup table
(2) but not good. Why? Previous we manage to align lookup input calculated inside circuit and lookup input in rw table one by one, so comparing one by one is ok, and the error log contains more information for debugging. But in the new return opcode here, a lot of Memory rw is used to carry the deployed contract from bus-mapping to circuit side while these Memory rws are not used in evm circuit, they are only used to pass data, it is different from previous opcodes design.   So @z2trillion previously every rw table entry matches a rw lookup inside evm circuit, while the new return opcode has some rw table entry not needed by circuit. Is there other appoaches? (I am not sure whether these memory Rw are needed in state circuit, if not, maybe we can add use some other new field to pass data from busmapping to zkevm circuit rather than a lot of Memory Rws)

Reply:
The modification is needed because the copy table lookup does not occur at the "end" of the bus mapping for this gadget. 

Each `CopyEvent` can result in rw table entries (see https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/bus-mapping/src/circuit_input_builder/execution.rs#L244 for exactly how many). We need to check the rw consistency of these in the state circuit, or we could wrongly prove arbitrary call return values.

For previously existing gadgets, we were not getting any warnings from this because the copy rw entries occured at the end of the rw entries for the gadget, so the `assigned_rw_values` iterator would terminate.

I should be able to modify this to get back all the debugging information that was previously available.